src/VecSim/spaces/IP/IP_SSE_FP64.cpp

```diff
@@ -0,0 +1,138 @@
+#include "IP_SSE.h"
+#include "IP.h"
+#include "VecSim/spaces/space_includes.h"
+
+double FP64_InnerProductSIMD8Ext_SSE_impl(const void *pVect1v, const void *pVect2v, size_t qty) {
+
+    double *pVect1 = (double *)pVect1v;
+    double *pVect2 = (double *)pVect2v;
+
+    const double *pEnd1 = pVect1 + qty;
+
+    __m128d v1, v2;
+    __m128d sum_prod = _mm_set1_pd(0);
+
+    // In each iteration we calculate 8 doubles = 512 bits in total.
+    while (pVect1 < pEnd1) {
+        v1 = _mm_loadu_pd(pVect1);
+        pVect1 += 2;
+        v2 = _mm_loadu_pd(pVect2);
+        pVect2 += 2;
+        sum_prod = _mm_add_pd(sum_prod, _mm_mul_pd(v1, v2));
+
+        v1 = _mm_loadu_pd(pVect1);
+        pVect1 += 2;
+        v2 = _mm_loadu_pd(pVect2);
+        pVect2 += 2;
+        sum_prod = _mm_add_pd(sum_prod, _mm_mul_pd(v1, v2));
+
+        v1 = _mm_loadu_pd(pVect1);
+        pVect1 += 2;
+        v2 = _mm_loadu_pd(pVect2);
+        pVect2 += 2;
+        sum_prod = _mm_add_pd(sum_prod, _mm_mul_pd(v1, v2));
+
+        v1 = _mm_loadu_pd(pVect1);
+        pVect1 += 2;
+        v2 = _mm_loadu_pd(pVect2);
+        pVect2 += 2;
+        sum_prod = _mm_add_pd(sum_prod, _mm_mul_pd(v1, v2));
+    }
+
+    double PORTABLE_ALIGN16 TmpRes[2];
+    _mm_store_pd(TmpRes, sum_prod);
+
+    return TmpRes[0] + TmpRes[1];
+}
+
+double FP64_InnerProductSIMD8Ext_SSE(const void *pVect1v, const void *pVect2v, size_t qty) {
+    return 1.0 - FP64_InnerProductSIMD8Ext_SSE_impl(pVect1v, pVect2v, qty);
+}
+
+double FP64_InnerProductSIMD8ExtResiduals_SSE(const void *pVect1v, const void *pVect2v,
+                                              size_t qty) {
+    size_t qty8 = qty >> 2 << 2;
```

Comment:
```suggestion
    size_t qty8 = qty >> 3 << 3;
```
How did we miss this one in the tests?

Reply:
Good catch
We test this one with dim = 17, so taking off %4 leaves us with 16 floats to calculate with FP64_InnerProductSIMD8Ext. Since 16 is also divisible by 8 it worked :sweat_smile: 
Ill change the dimension to 13.